### PR TITLE
Multipart

### PR DESCRIPTION
**Pull request by yohanboniface**

### CODE REVIEW

File: docs/reference.md [hunk 1]
```diff
@@ -95,6 +101,29 @@ response.json = {'some': 'dict'}
 response.json = [{'some': 'dict'}, {'another': 'one'}]
 ```
 
+### Multipart
+
+Responsible of the parsing of `request.body`.
```
Comment (yohanboniface) [inline docs/reference.md]: "... of multipart encoded request.body" ?

File: docs/reference.md [hunk 2]
```diff
@@ -95,6 +101,29 @@ response.json = {'some': 'dict'}
...
+
+#### Methods
+
+- **parse(content_type: str)**: returns a tuple
```
Comment (yohanboniface) [inline docs/reference.md]: Not sure the name is that good: it should imho reflects the two steps workflow, i.e. "init" then "feed_data" as many times as needed. So maybe `init`.
Or we use the `__init__` (which does not return anything then), and we retrieve form and files `self._form = parser.form`.

File: docs/reference.md [hunk 3]
```diff
@@ -95,6 +101,29 @@ response.json = {'some': 'dict'}
...
+  ([Form](#form) instance, [Files](#files) instance) filled with data
+  from `feed_data`
+- **feed_data(data: bytes)**: incrementally fills [Form](#form) and
+  [Files](#files) objects with bytes from the body
+
+
+### Form
+
+Allow to access casted POST parameters from `request.body`.
+
```
Comment (yohanboniface) [inline docs/reference.md]: We should clarify that Form inherits from Query, with all the goodies.

File: docs/reference.md [hunk 4]
```diff
@@ -95,6 +101,29 @@ response.json = {'some': 'dict'}
...
+
+### Files
+
+Allow to access POSTed files from `request.body`.
```
Comment (yohanboniface) [inline docs/reference.md]: We should document `get` and `list` method (from Multidict).

File: roll/__init__.py [hunk 5]
```diff
@@ -98,7 +109,80 @@ def float(self, key: str, default=...):
...
+    def parse(self, content_type: str):
+        self._parser = Parser(self, content_type.encode())
+        self.form = self.app.Form()
+        self.files = self.app.Files()
+        return self.form, self.files
+
+    def feed_data(self, data: bytes):
+        self._parser.feed_data(data)
+
+    def on_body_begin(self):
```
Comment (yohanboniface) [inline roll/__init__.py]: We may make all those event listener optional, multifruits side.

Comment (yohanboniface) [inline roll/__init__.py]: We may make all those event listener optional, multifruits side.

File: roll/__init__.py [hunk 6]
```diff
@@ -132,6 +218,41 @@ def query(self):
...
+        if self._files is None:
+            self._parse_multipart()
+        return self._files
+
+    @property
+    def json(self):
+        try:
+            return json.loads(self.body.decode())
+        except (UnicodeDecodeError, JSONDecodeError):
+            return {}
```
Comment (yohanboniface) [inline roll/__init__.py]: Maybe `None` would be a better default?

File: roll/testing.py [hunk 7]
```diff
@@ -25,6 +79,30 @@ class Client:
     def __init__(self, app):
         self.app = app
 
+    def handle_files(self, kwargs):
```
Comment (yohanboniface) [inline roll/testing.py]: Not related to this PR, but let's have a dedicated doc page (how-to) for "testing".

File: roll/__init__.py (hunk 6 above)
Comment (davidbgk) [reply roll/__init__.py]: It would raise an error in case of `request.json.get('foo')`.

File: docs/reference.md (hunk 4 above)
Comment (davidbgk) [reply docs/reference.md]: Done.

File: roll/testing.py (hunk 7 above)
Comment (davidbgk) [reply roll/testing.py]: Done!

File: roll/__init__.py [hunk 8]
```diff
@@ -132,6 +212,41 @@ def query(self):
...
+            if 'multipart/form-data' in self.content_type:
+                self._parse_multipart()
+            elif 'application/x-www-form-urlencoded' in self.content_type:
+                self._parse_urlencoded()
+        return self._form
+
+    @property
+    def files(self):
+        if self._files is None:
+            self._parse_multipart()
```
Comment (yohanboniface) [inline roll/__init__.py]: We should handle the case where it's not a multipart body.

File: tests/test_views.py [hunk 9]
```diff
@@ -50,7 +50,7 @@
 async def test_post_json(client, app):
 
     @app.route('/test', methods=['POST'])
-    async def get(req, resp):
+    async def post(req, resp):
```
Comment (yohanboniface) [inline tests/test_views.py]: Weird that `assert resp.status == HTTPStatus.OK` hasn't raise before :s

File: tests/test_request.py [hunk 10]
```diff
@@ -443,9 +475,8 @@ def protocol(app):
 
     @app.route('/test', methods=['POST'])
     async def post(req, resp):
-        assert req.json == {}
-        resp.body = b'done'
+        assert req.json
```
Comment (yohanboniface) [inline tests/test_request.py]: `raise` to make it clearer that we should not hit that piece of code?

Comment (davidbgk) [reply tests/test_request.py]: Actually, that's the instantiation that raises the error, maybe better to do `resp.body = req.json`?

**Pull request merged**